### PR TITLE
2021 Q2 updates (TEAL4, ASA URL, Program Sizes...)

### PR DESCRIPTION
**Pull request by jannotti**
These are new opcodes to support use-cases such as compiling to TEAL, or more complicated math in AMMs or token bridges.

### CODE REVIEW

File: dev/TEAL_opcodes.md [hunk 1]
```diff
@@ -425,6 +433,7 @@ Overflow is an error condition which halts execution and fails the transaction.
 | 53 | GlobalNumByteSlice | uint64 | Number of global state byteslices in ApplicationCall. LogicSigVersion >= 3. |
 | 54 | LocalNumUint | uint64 | Number of local state integers in ApplicationCall. LogicSigVersion >= 3. |
 | 55 | LocalNumByteSlice | uint64 | Number of local state byteslices in ApplicationCall. LogicSigVersion >= 3. |
+| 56 | AppProgramExtraPages | uint64 |  |
```
Comment (fabrice102) [inline dev/TEAL_opcodes.md:436]: Missing description

Comment (jannotti) [reply dev/TEAL_opcodes.md]: fixed

File: dev/ledger.md [hunk 2]
```diff
@@ -867,7 +875,9 @@ the transactions have nonzero "Group", compute the _TxGroup hash_ as follows:
 
 If the TxGroup hash of any transaction group in a block does not match the "Group" field of the transactions in that group (and that "Group" field is nonzero), then the block is invalid. Additionally, if a block contains a transaction group of more than $G_{max}$ transactions, the block is invalid.
 
-Beyond this check, each transaction in a group is evaluated separately and must be valid on its own, as described below in the [Validity and State Changes][Validity and State Changes] section. For example, a group containing a zero-fee transaction and a very-high-fee transaction would be rejected because the first transaction has fee less than $f_{\min}$, even if the average transaction fee of the group were above $f_{\min}$. As another example, an account with balance 50 could not spend 100 in transaction A and afterward receive 500 in transaction B, even if transactions A and B are in the same group, because transaction A would leave the account with a negative balance.
+If the sum of the fees paid by the transactions in a transaction group is less than $f_{\min}$ times the number of transactions in the group, then the block is invalid.
```
Comment (fabrice102) [inline dev/ledger.md:878]: What happens in case of congestion and fees are per byte?
I am not sure it's fully specified in the specs even for 1 transaction though.
This does not make the block invalid so it may be omitted from the specs (as all block proposers may not see the same congestion).
But it needs to be specified somewhere I think.

Comment (fabrice102) [inline dev/ledger.md:878]: In theory, this change of fees may cause some smart contracts to become insecure.
For example, a smart contract that does: 
> if the fee of the transaction is 0, then do something insecure, otherwise process normally

would now be insecure.
But I don't believe there is really a risk that such smart contracts exist.

Furthermore, I don't think the specs forces the min fee to be more than 0, so the smart contracts would not be fully secure in any implementation of Algorand's specs.

In conclusion, I don't think it is an issue.

Comment (fabrice102) [inline dev/ledger.md:878]: Currently pending rewards are committed any time the account's Algo balance is modified.
In particular, any time an account is a sender for a transaction, rewards are committed because paying the fee modifies the Algo balance.

Now that the fee can be 0 for a transaction, is it still the case?

File: dev/ledger.md [hunk 3]
```diff
@@ -1030,13 +1040,15 @@ point must be discarded and the entire transaction rejected.
...
-  local or global [TEAL Key/Value Store][TEAL Key/Value Stores] longer than 64
-  bytes, or writing any value to a key longer than 64 bytes, will likewise cause
-  the program to fail on the offending instruction.
+- During the execution of an `ApprovalProgram` or `ClearStateProgram`,
+  the application’s `LocalStateSchema` and `GlobalStateSchema` may
+  never be violated. The program's execution will fail on the first
+  instruction that would cause the relevant schema to be
+  violated. Writing a `Bytes` value to a local or global [TEAL
+  Key/Value Store][TEAL Key/Value Stores] such that the sum of the
+  lengths of the key and value in bytes exceeds 128, or writing any
```
Comment (fabrice102) [inline dev/ledger.md:1049]: Is 128 a consensus parameter?

Comment (fabrice102) [inline dev/ledger.md:1049]: Can applications with TEAL v2 or v3 benefit from the new rule that key size + value size must not exceed 128 bytes instead of key size and value size individually must not exceed 64 bytes?

Comment (fabrice102) [inline dev/ledger.md:1049]: Same question as above except for applications that were deployed before this consensus update.

**Review by fabrice102 — CHANGES_REQUESTED** [review]

File: dev/TEAL.md [hunk 4]
```diff
@@ -145,6 +149,42 @@ For two-argument ops, `A` is the previous element on the stack and `B` is the la
...
+as big-endian unsigned integers.  Returned values are the shortest
+byte-array that can represent the returned value.  For example, the
+zero value is the empty byte-array.
+
+Input lengths are limited to maximum length 64, which represents a 512
+bit unsigned integer.
+
+| Op | Description |
+| --- | --- |
+| `b+` | A plus B, where A and B are byte-arrays interpreted as big-endian unsigned integers |
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:162]: No.  There is an input size limitation (64 bytes), but *not* an output limitation (except the normal 4k limitation which will not come into play with these operations).  I that since we had enough room to represent the answer, there was no need to panic, even though you could not continue to operate on such a large number.  But you can certainly do other things with it.

File: dev/TEAL.md [hunk 5]
```diff
@@ -35,7 +35,7 @@ A program can either authorize some delegated action on a normal private key sig
...
 
-The TEAL bytecode plus the length of any Args must add up to less than 1000 bytes (consensus parameter LogicSigMaxSize). Each TEAL op has an associated cost estimate and the program cost estimate must total less than 20000 (consensus parameter LogicSigMaxCost). Most ops have an estimated cost of 1, but a few slow crypto ops are much higher.
+The TEAL bytecode plus the length of any Args must add up to less than 1000 bytes (consensus parameter LogicSigMaxSize). Each TEAL op has an associated cost and the program cost must total less than 20000 (consensus parameter LogicSigMaxCost). Most ops have a cost of 1, but a few slow crypto ops are much higher. Prior to v4, program costs was estimated as the static sum of all opcode costs in a program (ignoring conditionals that might skip some code). Beginning with v4, a program's cost is tracked dynamically, while being evaluated. If the program exceeds its budget, it fails.
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:38]: I'll make the change in the input file.

File: dev/TEAL.md [hunk 6]
```diff
@@ -88,11 +88,8 @@ An application transaction must indicate the action to be taken following the ex
...
-The instruction set was designed to execute calculator-like expressions.
-What might be a one line expression with various parenthesized clauses should be efficiently representable in TEAL.
 
-Looping is not possible, by design, to ensure predictably fast execution.
-There is a branch instruction (`bnz`, branch if not zero) which allows forward branching only so that some code may be skipped.
+Many instructions accept values to designate Accounts, Assets, or Applications. Beginning with TEAL v4, these values may always be given as an _offset_ in the corresponding Txn fields (Txn.Accounts, Txn.ForeignAssets, Txn.ForeignApps) _or_ as the value itself (a bytes address for Accounts, or a uint64 id). The values, however, must still be present in the Txn fields. Before TEAL v4, most opcodes required the use of an offset, except for reading account local values of assets or applications, which accepted the IDs directly and did not require the ID to be present in they corresponding _Foreign_ array.  See individual opcodes for details. In the case of account offsets or application offsets, 0 is specially defined to Txn.Sender or the ID of the current application, respectively.
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:92]: I'll make the change in the input file.

File: dev/TEAL.md [hunk 7]
```diff
@@ -124,6 +121,11 @@ For two-argument ops, `A` is the previous element on the stack and `B` is the la
 | `>=` | A greater than or equal to B => {0 or 1} |
 | `&&` | A is not zero and B is not zero => {0 or 1} |
 | `\|\|` | A is not zero or B is not zero => {0 or 1} |
+| `shl` | A times 2^B, modulo 2^64 |
+| `shr` | A divided by 2^B |
+| `sqrt` | The largest integer X such that X^2 <= A |
+| `bitlen` | The index of the highest bit in A. If A is a byte-array, it is interpreted as a big-endian unsigned integer |
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:127]: Since bitlen starts with bitlen(0) == 0, I decided the word "index" in confusing.  So how about: "The highest set bit in A. If A is a byte-array, it is interpreted as a big-endian unsigned integer. bitlen of 0 is 0, bitlen of 8 is 4"

File: dev/TEAL.md [hunk 8]
```diff
@@ -167,6 +207,7 @@ Some of these have immediate data in the byte or bytes after the opcode.
 | `bytec_2` | push constant 2 from bytecblock to stack |
 | `bytec_3` | push constant 3 from bytecblock to stack |
 | `pushbytes bytes` | push the following program bytes to the stack |
+| `bzero` | push a byte-array of length A, containing all zero bytes |
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:210]: Thanks, I got b~ as well.

File: dev/ledger.md (hunk 2 above)
Comment (jannotti) [reply dev/ledger.md:878]: Excellent question.  I'll investigate.  Is this spec'd anywhere?

File: dev/ledger.md (hunk 3 above)
Comment (jannotti) [reply dev/ledger.md:1049]: Yes.  Should I name it here?  We don't seem to do that in ledger.md, where there are several such (maximum schema sizes, program costs, app argument count and size).

Comment (jannotti) [reply dev/ledger.md:1049]: Yes.  I suppose there is, again, the theoretical issue that an old smart contract could have been *intending* to panic if its value exceeded 64.

Comment (jannotti) [reply dev/ledger.md:1049]: Yes. And same concern.

File: dev/TEAL.md [hunk 9]
```diff
@@ -44,19 +44,19 @@ Starting from version 2 TEAL evaluator can run programs in two modes:
 2. Application run (stateful)
 
 Differences between modes include:
-1. Max program length (consensus parameters LogicSigMaxSize, MaxApprovalProgramLen and MaxClearStateProgramLen)
+1. Max program length (consensus parameters LogicSigMaxSize, MaxAppProgramLen & MaxExtraAppProgramPages)
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:47]: Yes, that's allowed.

[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:47]: Yes, and no.  The real bounds are derived from MaxAppTotalProgramLen (which is one "page" - it's called "Total" to reflect that it is the sum of Approval and ClearState) and MaxExtraAppProgramPages (which is the number of such program lengths).  So the actual test is:
```
		lap := len(tx.ApprovalProgram)
		lcs := len(tx.ClearStateProgram)
		pages := int(1 + tx.ExtraProgramPages)
		if lap+lcs > pages*proto.MaxAppTotalProgramLen {
			return fmt.Errorf("app programs too long. max total len %d bytes", pages*proto.MaxAppTotalProgramLen)
		}
```
There are also individual tests against MaxAppProgramLen, but those are essentially no-ops in the new protocol level because MaxAppProgramLen  and MaxAppTotalProgramLen are now the same.

I'll make the correction.

File: dev/TEAL_opcodes.md [hunk 10]
```diff
@@ -522,20 +531,64 @@ for notes on transaction fields available, see `txn`. If top of stack is _i_, `g
...
+## gload t i
+
+- Opcode: 0x3a {uint8 transaction group index} {uint8 position in scratch space to load from}
+- Pops: _None_
+- Pushes: any
+- push Ith scratch space index of the Tth transaction in the current group
+- LogicSigVersion >= 4
+- Mode: Application
+
+The `gload` opcode can only access scratch spaces of previous app calls contained in the current group.
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL_opcodes.md:543]: It gives uint64 0, which is the same behaviour as `load`

File: dev/TEAL_opcodes.md [hunk 11]
```diff
@@ -680,44 +733,46 @@ bit indexing begins with low-order bits in integers. Setting bit 4 to 1 on the i
...
 - Opcode: 0x60
-- Pops: *... stack*, uint64
+- Pops: *... stack*, any
 - Pushes: uint64
-- get balance for the requested account specified by Txn.Accounts[A] in microalgos. A is specified as an account index in the Accounts field of the ApplicationCall transaction, zero index means the sender. The balance is observed after the effects of previous transactions in the group, and after the fee for the current transaction is deducted.
+- get balance account A, in microalgos. The balance is observed after the effects of previous transactions in the group, and after the fee for the current transaction is deducted.
 - LogicSigVersion >= 2
 - Mode: Application
 
+params: Txn.Accounts offset (or, since v4, an account address that appears in Txn.Accounts or is Txn.Sender). Return: value.
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL_opcodes.md:742]: I like your wording, but I'll wait for a final consensus with Shai, since there are a bunch to change.

File: dev/TEAL_opcodes.md [hunk 12]
```diff
@@ -851,10 +908,223 @@ params: txn.ForeignAssets offset. Return: did_exist flag (1 if exist and 0 other
...
+- Pushes: uint64
+- A divided by 2^B
+- LogicSigVersion >= 4
+
+## sqrt
+
+- Opcode: 0x92
+- Pops: *... stack*, uint64
+- Pushes: uint64
+- The largest integer X such that X^2 <= A
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL_opcodes.md:964]: I'll fix, maybe using BASE.

File: dev/TEAL_opcodes.md [hunk 13]
```diff
@@ -851,10 +908,223 @@ params: txn.ForeignAssets offset. Return: did_exist flag (1 if exist and 0 other
...
+
+## bitlen
+
+- Opcode: 0x93
+- Pops: *... stack*, any
+- Pushes: uint64
+- The index of the highest bit in A. If A is a byte-array, it is interpreted as a big-endian unsigned integer
+- LogicSigVersion >= 4
+
+bitlen interprets arrays as big-endian integers, unlike setbit/getbit
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL_opcodes.md:976]: Changed.

File: dev/TEAL_opcodes.md [hunk 14]
```diff
@@ -851,10 +908,223 @@ params: txn.ForeignAssets offset. Return: did_exist flag (1 if exist and 0 other
...
+
+## b+
+
+- Opcode: 0xa0
+- Pops: *... stack*, {[]byte A}, {[]byte B}
+- Pushes: []byte
+- A plus B, where A and B are byte-arrays interpreted as big-endian unsigned integers
+- **Cost**: 10
+- LogicSigVersion >= 4
+
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL_opcodes.md:1003]: It is not an overflow.

File: dev/ledger.md [hunk 15]
```diff
@@ -325,8 +325,7 @@ parameters_, which can be encoded as a msgpack struct:
   global state associated with this application. This field is encoded with
   msgpack field `approv`.
 
-  This field must not exceed 1024 bytes. The cost of the program as determined
-  by the Stateful TEAL `Check` function must not exceed 700.
+  The cost of the program during execution must not exceed 700.
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/ledger.md:328]: Yes.  I'll fix.

File: dev/ledger.md [hunk 16]
```diff
@@ -325,8 +325,7 @@ parameters_, which can be encoded as a msgpack struct:
...
 
 - A mutable Stateful TEAL "Clear State" program (`ClearStateProgram`), executed
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/ledger.md:330]: Good question, I'll investigate the state question. For `gload` I believe the answer is that they are readable.

File: dev/ledger.md (hunk 2 above)
Comment (jannotti) [reply dev/ledger.md:878]: I don't think we can include that here, since it's not protocol spec.  I'm open to where it could best be documented.

File: dev/TEAL.md [hunk 17]
```diff
@@ -35,7 +35,7 @@ A program can either authorize some delegated action on a normal private key sig
 * If the account has signed the program (an ed25519 signature on "Program" concatenated with the program bytes) then if the program returns true the transaction is authorized as if the account had signed it. This allows an account to hand out a signed program so that other users can carry out delegated actions which are approved by the program.
```
[the remark this answers is not in the document]
Comment (shaih) [reply dev/TEAL.md:35]: Perhaps change "Execution environment" to "Execution environment in stateless mode" and start the section by "In stateless mode, TEAL runs in Algorand nodes as part of testing a proposed transaction to see if it is valid and authorized to be committed into a block."

File: dev/TEAL.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (shaih) [reply dev/TEAL.md:164]: So maybe say it. Add at the end of lines 160: "Output values are not length-limited, and will be stored on the stack even if they exceed 64 bytes"

File: dev/TEAL.md [hunk 18]
```diff
@@ -145,6 +151,42 @@ For two-argument ops, `A` is the previous element on the stack and `B` is the la
...
+| `b>=` | A is greater than or equal to B, where A and B are byte-arrays interpreted as big-endian unsigned integers => { 0 or 1} |
+| `b==` | A is equals to B, where A and B are byte-arrays interpreted as big-endian unsigned integers => { 0 or 1} |
+| `b!=` | A is not equal to B, where A and B are byte-arrays interpreted as big-endian unsigned integers => { 0 or 1} |
+| `b%` | A modulo B, where A and B are byte-arrays interpreted as big-endian unsigned integers. Panic if B is zero. |
+
+These opcodes operate on the bits of byte-array values.  The shorter
+array is interpeted as though left padded with zeros until it is the
+same length as the other input.  The returned values are the same
+length as the longest input.  Therefore, unlike array arithmetic,
+these results may contain leading zero bytes.
```
Comment (shaih) [inline dev/TEAL.md:180]: Doesn't this contradict the previous discussion? Can't the result be longer than both input values?

File: dev/TEAL.md [hunk 19]
```diff
@@ -305,24 +353,26 @@ Asset fields include `AssetHolding` and `AssetParam` fields that are used in `as
...
-| `app_opted_in` | check if account specified by Txn.Accounts[A] opted in for the application B => {0 or 1} |
-| `app_local_get` | read from account specified by Txn.Accounts[A] from local state of the current application key B => value |
-| `app_local_get_ex` | read from account specified by Txn.Accounts[A] from local state of the application B key C => [*... stack*, value, 0 or 1] |
+| `balance` | get balance account A, in microalgos. The balance is observed after the effects of previous transactions in the group, and after the fee for the current transaction is deducted. |
+| `min_balance` | get minimum required balance account A, in microalgos. Required balance is affected by [ASA](https://developer.algorand.org/docs/features/asa/#assets-overview) and [App](https://developer.algorand.org/docs/features/asc1/stateful/#minimum-balance-requirement-for-a-smart-contract) usage. When creating or opting into an app, the minimum balance grows before the app code runs, therefore the increase is visible there. When deleting or closing out, the minimum balance decreases after the app executes. |
```
Comment (shaih) [inline dev/TEAL.md:364]: "get minimum required balance *for* account A"

File: dev/TEAL_opcodes.md [hunk 20]
```diff
@@ -522,20 +531,64 @@ for notes on transaction fields available, see `txn`. If top of stack is _i_, `g
...
+- Mode: Application
+
+The `gloads` opcode can only access scratch spaces of previous app calls contained in the current group.
+
+## gaid t
+
+- Opcode: 0x3c
+- Pops: _None_
+- Pushes: uint64
+- push the ID of the asset or application created in the Tth transaction of the current group
```
Comment (shaih) [inline dev/TEAL_opcodes.md:561]: What happens if that transaction did not create any asset/application?

File: dev/TEAL_opcodes.md [hunk 21]
```diff
@@ -522,20 +531,64 @@ for notes on transaction fields available, see `txn`. If top of stack is _i_, `g
...
+- push the ID of the asset or application created in the Ath transaction of the current group
+- LogicSigVersion >= 4
+- Mode: Application
+
+The `gaids` opcode can only access the ID of assets or applications created by previous txns in the current group.
+
 ## bnz target
 
-- Opcode: 0x40 {0..0x7fff forward branch offset, big endian}
+- Opcode: 0x40 {int16 branch offset, big endian. (negative offsets are illegal before v4)}
```
Comment (shaih) [inline dev/TEAL_opcodes.md:580]: Maybe rephrase the parenthesized sentence: "(negative offsets are only allowed starting v4, they are illegal in prior versions)"

File: dev/TEAL_opcodes.md [hunk 22]
```diff
@@ -522,20 +531,64 @@ for notes on transaction fields available, see `txn`. If top of stack is _i_, `g
...
+The `bnz` instruction opcode 0x40 is followed by two immediate data bytes which are a high byte first and low byte second which together form a 16 bit offset which the instruction may branch to. For a bnz instruction at `pc`, if the last element of the stack is not zero then branch to instruction at `pc + 3 + N`, else proceed to next instruction at `pc + 3`. Branch targets must be well aligned instructions. (e.g. Branching to the second byte of a 2 byte op will be rejected.) Branch offsets are limited to forward branches only, 0-0x7fff until v4. v4 treats offset as a signed 16 bit integer allowing for backward branches and looping.
```
Comment (shaih) [inline dev/TEAL_opcodes.md:585]: Replace the last two sentences: "Starting at v4, the offset is treated as a signed 16 bit integer allowing for backward branches and looping. In prior version (v1 to v3), branch offsets are limited to forward branches only, 0-0x7fff."

File: dev/TEAL_opcodes.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (shaih) [reply dev/TEAL_opcodes.md:742]: I like Fabrice's wording

File: dev/ledger.md (hunk 3 above)
Comment (shaih) [reply dev/ledger.md:1053]: I don't follow. Does that mean that stateful TEAL programs are not versioned? So a stateful TEAL v2 program that would be called after the upgrade will run according to TEALv4 rules?

File: dev/ledger.md [hunk 23]
```diff
@@ -1045,14 +1061,20 @@ point must be discarded and the entire transaction rejected.
...
-  transaction’s `Accounts` field. An attempt to read local state from any other
-  account will cause program execution to fail.
-- Algo balances and asset balances may be read for the sender's account or for
-  any account referenced by an address listed in the transaction's `Accounts`
-  field. An attempt to read an Algo balance or asset balance for any other
-  account will cause program execution to fail.
+- Local state may be read for any opted-in application present in the
+  sender’s account data, or in the account data for any address listed
+  in the transaction’s `Accounts` field. An attempt to read local
+  state from any other tea account will cause program execution to
```
Comment (shaih) [inline dev/ledger.md:1067]: "tea account"?

File: dev/TEAL.md (hunk 18 above)
Comment (jannotti) [reply dev/TEAL.md:180]: The distinction is between the bytes math and the bytes bitwise operators. Because of the way we expect them to be used, we felt the bitwise operators would be easier to understand and use if they maintained their lengths.  When reading TEAL.md as a whole, I think this is clear because they are in different paragraphs.  It does make me realize I need to be explicit about the bytes comparison operators, however, as they return normal uint64 {0, 1}.  I need to make it clear they are *not* returning bytes that represent those numbers.

File: dev/TEAL_opcodes.md (hunk 20 above)
Comment (jannotti) [reply dev/TEAL_opcodes.md:561]: panic.  I'll update.

File: dev/TEAL_opcodes.md (hunk 21 above)
Comment (jannotti) [reply dev/TEAL_opcodes.md:580]: I removed that explanation there, as that should really be a concise explanation of the encoding of the immediate.  The `b` instruction explains the versioning rules, and I've just updated it to be a bit clearer (to be pushed very soon)

File: dev/TEAL_opcodes.md (hunk 22 above)
Comment (jannotti) [reply dev/TEAL_opcodes.md:585]: This was what I just changed.  I'll use your language.

File: dev/ledger.md (hunk 23 above)
Comment (jannotti) [reply dev/ledger.md:1067]: removed "tea"

File: dev/TEAL.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:164]: Input lengths are limited to a maximum length 64 bytes, which
represents a 512 bit unsigned integer. Output lengths are not
explicitly restricted, though only `b*` and `b+` can produce a larger
output than their inputs, so there is an implicit length limit of 128
bytes on outputs.

File: dev/TEAL_opcodes.md [hunk 24]
```diff
@@ -851,10 +908,223 @@ params: txn.ForeignAssets offset. Return: did_exist flag (1 if exist and 0 other
...
+
+## callsub target
+
+- Opcode: 0x88
+- Pops: _None_
+- Pushes: _None_
+- branch unconditionally to TARGET, saving the next instruction on the call stack
+- LogicSigVersion >= 4
+
+The call stack is separate from the data stack. Only `callsub` and `retsub` manipulate it.`
```
Comment (barnjamin) [inline dev/TEAL_opcodes.md:931]: backtick at the end of the description here and on retsub intentional?

File: dev/TEAL.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:45]: The LogicSig specific stuff starts even earlier.  I'm committing slightly more extensive changes to introduce Applications earlier.

File: dev/ledger.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/ledger.md:332]: Somewhat surprising to me, my reading of the code implies that changes made during a "failed" Clear State are *not* committed.  I will ask around to be sure.  This does seem like it should be better spec'd.

File: dev/TEAL_opcodes.md (hunk 24 above)
Comment (jannotti) [reply dev/TEAL_opcodes.md:931]: Thanks, I removed them once, but in a stranded PR that didn't go in.

File: dev/ledger.md (hunk 3 above)
Comment (jannotti) [reply dev/ledger.md:1053]: TEAL programs are versioned, but these things are done outside of the TEAL environment.  TEAL doesn't force values to be less than 64 bytes, the rules for writing ApplyData does.  Similarly, TEAL doesn't set program sizes, the transaction processing rules do.

File: dev/TEAL.md [hunk 25]
```diff
@@ -242,6 +292,7 @@ Some of these have immediate data in the byte or bytes after the opcode.
 | 53 | GlobalNumByteSlice | uint64 | Number of global state byteslices in ApplicationCall. LogicSigVersion >= 3. |
 | 54 | LocalNumUint | uint64 | Number of local state integers in ApplicationCall. LogicSigVersion >= 3. |
 | 55 | LocalNumByteSlice | uint64 | Number of local state byteslices in ApplicationCall. LogicSigVersion >= 3. |
+| 56 | ExtraProgramPages | uint64 | Number of additional pages for each of the application's approval and clear state programs. An ExtraProgramPages of 1 means 2048 more total bytes, or 1024 for each program. LogicSigVersion >= 4. |
```
Comment (jeapostrophe) [inline dev/TEAL.md:295]: It's too hard to make the sizes of the two programs independent? I've never written a ClearStateProgram more than ~10 bytes... it hurts me to pay for 1024 for both when I really only care about the ApprovalProgram

Comment (jannotti) [reply dev/TEAL.md:295]: Yes, we did both.  ExtraProgramPages gives you more space.  The changes to MaxAppProgramLen and introduction of MaxAppTotalProgramLen make it so that the bytes are shared freely between the two programs.

Comment (jeapostrophe) [reply dev/TEAL.md:295]: Excellent, I didn't catch the repercussions

File: dev/ledger.md (hunk 2 above)
Comment (fabrice102) [reply dev/ledger.md:882]: Unfortunately, I don't think it's spec'd anywhere.

File: dev/ledger.md (hunk 3 above)
Comment (fabrice102) [reply dev/ledger.md:1053]: I think it would be best to always name them. But no need to do it in this commit. This can be a later commit cleaning names.

Comment (fabrice102) [reply dev/ledger.md:1053]: Yes, I think it's fine. In theory, there is a security issue.
In practice, it is extremely unlikely someone did that.

File: dev/TEAL_opcodes.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (fabrice102) [reply dev/TEAL_opcodes.md:543]: Should we add some text?
That happens if the Tth transaction does not exist or is not an app call?
Is it also loading uint64 0?

File: dev/TEAL_opcodes.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (fabrice102) [reply dev/TEAL_opcodes.md]: `B` is usually the second argument, I think it's even better to use `Y` which looks more like an output and cannot be confused as an input.
What do you think?

File: dev/ledger.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (fabrice102) [reply dev/ledger.md:332]: I think that makes sense they are not committed. But it needs to be spec'd.

File: dev/TEAL.md [hunk 26]
```diff
@@ -22,9 +21,20 @@ The maximum stack depth is currently 1000.
 
 In addition to the stack there are 256 positions of scratch space, also uint64-bytes union values, accessed by the `load` and `store` ops moving data from or to scratch space, respectively.
 
-## Execution Environment
+## Execution modes
```
Comment (fabrice102) [inline dev/TEAL.md:24]: "modes" -> "Modes"

File: dev/TEAL.md [hunk 27]
```diff
@@ -145,6 +150,46 @@ For two-argument ops, `A` is the previous element on the stack and `B` is the la
 | `substring s e` | pop a byte-array A. For immediate values in 0..255 S and E: extract a range of bytes from A starting at S up to but not including E, push the substring result. If E < S, or either is larger than the array length, the program fails |
 | `substring3` | pop a byte-array A and two integers B and C. Extract a range of bytes from A starting at B up to but not including C, push the substring result. If C < B, or either is larger than the array length, the program fails |
 
+These opcodes take byte-array values that are interpreted as
+big-endian unsigned integers.  For mathematical opeartors, the
```
Comment (fabrice102) [inline dev/TEAL.md:154]:
```suggestion
big-endian unsigned integers.  For mathematical operators, the
```

File: dev/TEAL_opcodes.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL_opcodes.md:543]: those panic.  I'll add detail.

**Review by fabrice102 — CHANGES_REQUESTED** [review]

File: dev/TEAL.md (hunk 18 above)
Comment (shaih) [reply dev/TEAL.md:183]: oh, I was reading "these opcodes" as talking about the ones above, but it means the ones below. Maybe "the following opcodes"? (Also in line 153)

File: dev/TEAL_opcodes.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL_opcodes.md:742]: This worked well for opcodes with a single parameter (balance, min_balance, asset_params_get). However, for opcodes that take multiple parameters, the repetition of all parameters made them hard to follow. The parenthetical "or" clauses seemed to work better, as they kept the parameters in order.  Since I believe both are correct, I'd like to leave those alone. If anything, perhaps saying these things two different ways for the different opcodes might make it clearer.  Maybe?  What do you think?

File: dev/TEAL.md [hunk 28]
```diff
@@ -22,9 +21,20 @@ The maximum stack depth is currently 1000.
 
 In addition to the stack there are 256 positions of scratch space, also uint64-bytes union values, accessed by the `load` and `store` ops moving data from or to scratch space, respectively.
 
-## Execution Environment
+## Execution Modes
+
+Starting from version 2 TEAL evaluator can run programs in two modes:
+1. LogigSig (stateless)
```
Comment (shaih) [inline dev/TEAL.md:27]:
```suggestion
1. LogicSig (stateless)
```

File: dev/TEAL.md [hunk 29]
```diff
@@ -176,11 +222,15 @@ Some of these have immediate data in the byte or bytes after the opcode.
...
 | `txna f i` | push Ith value of the array field F of the current transaction |
 | `gtxna t f i` | push Ith value of the array field F from the Tth transaction in the current group |
-| `gtxns f` | push field F of the Ath transaction in the current group |
-| `gtxnsa f i` | push Ith value of the array field F from the Ath transaction in the current group |
+| `gtxns f` | push field F of the Xth transaction in the current group |
+| `gtxnsa f i` | push Ith value of the array field F from the Xth transaction in the current group |
 | `global f` | push value from globals to stack |
 | `load i` | copy a value from scratch space to the stack |
 | `store i` | pop a value from the stack and store to scratch space |
+| `gload t i` | push Ith scratch space index of the Tth transaction in the current group |
```
Comment (jeapostrophe) [inline dev/TEAL.md:230]: These instructions are useful, but to be completely clear, it is very important that the scratch space also be accessible after the txn group is completed when we look up information about it in the indexer. This is so that we can have the smart contract have "output" in the sense of return arguments from method calls (including pure ones) and so that we can have "event logs" (a la Ethereum) by defining an ABI for the format of the final scratch space

Comment (jannotti) [reply dev/TEAL.md:230]: That is probably not the plan, as it is potentially quite large (256*4k).  We intend to have a `announce` operator or similar, but it will be more intentional, so that we aren't committing to have the scratch space kept around for a contract that might not even care.

File: dev/TEAL.md (hunk 25 above)
Comment (fabrice102) [inline dev/TEAL.md:295]: Might be a bit consufing "or 1024 for each program", but let's keep it this way.

**Review by fabrice102 — APPROVED** [review]
